### ISSUE TITLE
VS Code is sending us incorrect completionItem/resolve data in CompletionItem.kind

### ISSUE THREAD

**Issue by sean-mcmanus** [Microsoft/vscode-languageserver-node#438]
Our C/C++ extension is adding completionItem_resolve support, but even though the initial textDocument_completion we send to VS Code has the correct CompletionItem.kind data, the data we get back in the completionItem_resolve has the incorrect type, which causes all the icons to become incorrect as the user scrolls through the completion list. We're using the 3.5.1 vscode-languageclient.

![image](https://user-images.githubusercontent.com/19859882/49845540-f3ecfe80-fd7b-11e8-8472-639b6c4f0b83.png)

The data we send has a "kind" value of 21, but we get back a "kind" value of 1:
```
{
    "jsonrpc": "2.0",
    "id": 37,
    "result": {
        "isIncomplete": false,
        "items": [
            {
                "label": "macro1",
                "kind": 21,
                "detail": "#define macro1 ",
                "sortText": "1macro1",
                "textEdit": {
                    "range": {
                        "start": {
                            "line": 5,
                            "character": 4
                        },
                        "end": {
                            "line": 5,
                            "character": 10
                        }
                    },
                    "newText": "macro1"
                },
                "data": {
                    "file": "...",
                    "symbolPosition": {
                        "line": 3,
                        "character": 12
                    }
                }
            },
```
```
{
    "jsonrpc": "2.0",
    "id": 38,
    "method": "completionItem/resolve",
    "params": {
        "label": "macro1",
        "detail": "#define macro1 ",
        "insertTextFormat": 1,
        "textEdit": {
            "newText": "macro1",
            "range": {
                "start": {
                    "line": 5,
                    "character": 4
                },
                "end": {
                    "line": 5,
                    "character": 10
                }
            }
        },
        "kind": 1,
        "sortText": "1macro1",
        "data": {
            "file": "...",
            "symbolPosition": {
                "line": 3,
                "character": 12
            }
        }
    }
}
```

This bug only repros with the "new" CompletionItemKind values of 19-25: Folder, EnumMember, Constant, Struct, Event, Operator, and TypeParameter.

**Comment by rcjsuen** [comment]
If you look at the `release/3.5.1` tag, which appears to be pointing at b0635ac289afbb6c7d9e29d28d4e568c0b5bccdc, it seems that the `CompletionItemKind`s you are trying to use do not exist.

https://github.com/Microsoft/vscode-languageserver-node/blob/b0635ac289afbb6c7d9e29d28d4e568c0b5bccdc/types/src/main.ts#L659-L683

I suggest you use a newer version of `vscode-languageclient`. @dbaeumer Do you agree with this assessment?

**Comment by sean-mcmanus** [comment]
Okay, thanks for your help. That caught us by surprise because everything was working until we added completionItem_resolve support later on. We had some problem upgrading to 4.0/5.0 so as a workaround we have just added the CompletionItem.kind to the user "data" field and then re-set the kind to the correct value. So you can close this issue if you want.

**Comment by dbaeumer** [comment]
This is how it works in the latest release. The spec for 3.5.2 only allowed the listed items. We relaxed that in a new version and if clients support this then they agreed to preserve the kind even if it is unknown to them.

@sean-mcmanus I would highly recommend that you move to a newer version. We will very likely not do any fixing on 3.x versions. Why couldn't you upgrade?

**Comment by dbaeumer** [comment]
@rcjsuen thanks for the nice analysis of the problem.

**Comment by sean-mcmanus** [comment]
Bob didn't want to upgrade for some reason and we hit the breaking change with the provideCompletionItem, but I just fixed it, so I think we'll switch to 5.1.1. We can't get 5.2.0 or newer to work due to issue: 
```
node_modules/vscode-languageclient/lib/protocolConverter.d.ts(42,95): error TS2694: Namespace ''vscode'' has no exported member 'Declaration'.
child_process.js:644
    throw err;
    ^
```
I assume we need to upgrade our vscode to a newer engine version to fix that, which we'll try for our following release.

**Comment by dbaeumer** [comment]
@sean-mcmanus yes, the latest version of LSP client library requires VS Code 1.30 or higher. Reason being is that the LSP has GoTo declaration support. So you will be able to drop your custom support.
